libc-bottom-half/headers/public/wasi/descriptor_table.h

```diff
@@ -0,0 +1,121 @@
+#ifndef DESCRIPTOR_TABLE_H
+#define DESCRIPTOR_TABLE_H
+
+#include <wasi/preview2.h>
+
+typedef struct {
+} tcp_socket_state_unbound_t;
+typedef struct {
+} tcp_socket_state_bound_t;
+typedef struct {
+} tcp_socket_state_connecting_t;
+typedef struct {
+} tcp_socket_state_listening_t;
+
+typedef struct {
+	streams_own_input_stream_t input;
+	poll_own_pollable_t input_pollable;
+	streams_own_output_stream_t output;
+	poll_own_pollable_t output_pollable;
+} tcp_socket_state_connected_t;
+
+typedef struct {
+	network_error_code_t error_code;
+} tcp_socket_state_connect_failed_t;
+
+// This is a tagged union. When adding/removing/renaming cases, be sure to keep the tag and union definitions in sync.
+typedef struct {
+	enum {
+		TCP_SOCKET_STATE_UNBOUND,
+		TCP_SOCKET_STATE_BOUND,
+		TCP_SOCKET_STATE_CONNECTING,
+		TCP_SOCKET_STATE_CONNECTED,
+		TCP_SOCKET_STATE_CONNECT_FAILED,
+		TCP_SOCKET_STATE_LISTENING,
+	} tag;
+	union {
+		tcp_socket_state_unbound_t unbound;
+		tcp_socket_state_bound_t bound;
+		tcp_socket_state_connecting_t connecting;
+		tcp_socket_state_connected_t connected;
+		tcp_socket_state_connect_failed_t connect_failed;
+		tcp_socket_state_listening_t listening;
+	};
+} tcp_socket_state_t;
+
+typedef struct {
+	tcp_own_tcp_socket_t socket;
```

Comment:
Are these lower level types also still integers at the wasm module boundary?  i.e. are we still mapping integer "file descriptors" to interpreter "resource handles" here?    Or are we actually using unforgable externrefs at the core wasm level now?

Reply:
They're still integers, but they're unforgable in the sense that they're indices into host-managed, per-component tables, which means that if a component makes a call to another component, passing a resource handle, the host will interpose itself and look up the resource in the caller's table for that index, find (or create) the corresponding index in the callee's table, and pass that index to the callee.  Which I assume is also how externrefs work at some level, but I'm still learning about those :)

@alexcrichton can perhaps chime in here if my understanding is incorrect.

Reply:
Thanks that makes sense.  I know we are going to be moving to externref at some point but I've not been following too closely recently.

It might be worth mention somewhere (perhaps in the PR description) that the current "unforgeable resource handles" as still just integers and the core wasm/C/wasi-libc level, just like file descriptors.